File: 04-verbose-hello-world-srv/00-verbose-hello-world-srv-common/src/main/zava/com/github/jinahya/hello/misc/c02rfc862/Rfc862Tcp1Client.js
// #!/usr/bin/node

'use strict';

const assert = require('node:assert');
const crypto = require('node:crypto');
const net = require('node:net');

const IPv4 = true;
const HOST = IPv4 ? '127.0.0.1' : '::1';
if (IPv4) assert(net.isIPv4(HOST)); else assert(net.isIPv6(HOST));
const PORT = 50007;


function main() {
    const client = new net.Socket();
    client.on('connect', function () {
        console.log('connected to %s:%d/%s', client.remoteAddress, client.remotePort, client.remoteFamily);
        const hash = crypto.createHash('sha256');
        let bytes = Math.floor(Math.random() * 65536);
        console.log('sending %d byte(s)', bytes);
        const length = Math.floor(Math.random() * 8192) + 1;
        console.log('array.length: %d', length);
        let array = new Uint8Array(length);
        while (bytes > 0) {
            crypto.getRandomValues(array);
            if (array.length > bytes) {
                array = array.slice(0, bytes);
            }
            if (!client.write(array)) {
                client.once('drain');
            }
            hash.update(array);
            bytes -= array.length;
        }
        client.end();
        console.log('digest: %s', hash.digest('base64'));
    });
    client.on('data', function (data) {
    });
    client.on('drain', function () {
    });
    client.on('error', function (error) {
        console.log('error: %s', error);
    });
    client.on('end', function () {
        client.destroy();
    });
    client.on('close', function () {
    });
    client.connect({
        host: HOST,
        port: PORT
    });
}

if (require.main === module) {
    main();
}

File: 04-verbose-hello-world-srv/00-verbose-hello-world-srv-common/src/main/zava/com/github/jinahya/hello/misc/c02rfc862/Rfc862Tcp1Client.py
#!/usr/bin/python3

import hashlib
import random
import socket
import sys
from base64 import b64encode
from contextlib import closing

IPv4 = True
FAMILY = socket.AF_INET if IPv4 else socket.AF_INET6
HOST = '127.0.0.1' if IPv4 else "::1"
PORT = 50007


def hook(event, args):
    print(f'\t[audit]: {event} with args={args}')


sys.addaudithook(hook)


# https://stackoverflow.com/a/77426018/330457
def randomize(a):
    a[:] = random.randbytes(len(a))


with closing(socket.socket(FAMILY, socket.SOCK_STREAM)) as s:
    # -------------------------------------------------------------------- bind
    if bool(random.getrandbits(1)):
        s.bind((HOST, 0))
        print(f'(optionally) bound to {s.getsockname()}')
    # ----------------------------------------------------------------- connect
    s.connect((HOST, PORT))
    print(f'connected to {s.getpeername()}')
    bytes_ = random.randint(0, 65536)
    print(f'sending {bytes_} byte(s)')
    array = bytearray(random.randint(1024, 8192))
    print(f'array.length: {len(array)}')
    hash_ = hashlib.sha256()
    while bytes_ > 0:
        # ---------------------------------------------------------------- send
        randomize(array)
        if bytes_ < len(array):
            array = array[0:bytes_]
        w = s.send(array)
        hash_.update(array)
        bytes_ -= w
        # ---------------------------------------------------------------- recv
        bytes_received = s.recv(len(array))
    # ---------------------------------------------------------------- shutdown
    s.shutdown(socket.SHUT_WR)
    # ------------------------------------------------------------- read-to-eof
    while True:
        bytes_received = s.recv(len(array))
        if len(bytes_received) == 0:
            break
    # ------------------------------------------------------------ print-digest
    print(f'digest: {b64encode(hash_.digest()).decode()}')
